docs/design.md

```diff
@@ -0,0 +1,116 @@
+Design
+======
+
+The main goals of libnetwork are highlighted in the [roadmap](../ROADMAP.md).
+This document describes how libnetwork has been designed in order to acheive this.
+Requirements for individual releases can be found on the [Project Page](https://github.com/docker/libnetwork/wiki)
+
+## Legacy Docker Networking
+
+Prior to libnetwork a container's networking was handled in both Docker Engine and libcontainer.
+Docker Engine was responsible for providing the configuration of the container's networking stack.
+Libcontainer would then use this information to create the necessary networking devices and move them in to a network namespace.
+This namespace would then be used when the container is started.
+
+## The Container Network Model
+
+Libnetwork implements Container Network Model (CNM) which formalizes the steps required to provide networking for containers while providing an abstraction that can be used to support multiple network drivers. The CNM is built on 3 main components.
+
+**Sandbox**
+
+A Sandbox contains the configuration of a container's network stack.
+This includes management of the container's interfaces, routing table and DNS settings. 
+An implementation of a Sandbox could be a Linux Network Namespace, a FreeBSD Jail or other similar concept.
+A Sandbox may contain *many* endpoints from *multiple* networks
+
+**Endpoint**
+
+An Endpoint joins a Sandbox to a Network.
+An implementation of an Endpoint could be a `veth` pair, an Open vSwitch internal port or similar.
+An Endpoint can belong to *only one* network but may only belong to *one* Sandbox
+
+**Network**
+
+A Network is a group of Endpoints that are able to communicate with each-other directly.
+An implementation of a Network could be a Linux bridge, a VLAN etc...
+Networks consist of *many* endpoints
+
+## API
+
+Consumers of the CNM, like Docker for example, interact through the following APIs
+
+The `NetworkController` object is created to manage the allocation of Networks and the binding of these Networks to a specific Driver
+Once a Network is created, `network.CreateEndpoint` can be called to create a new Endpoint in a given network.
+When an Endpoint exists, it can be joined to a Sandbox using `endpoint.Join(id)`. If no Sandbox exists, one will be created, but if the Sandbox already exists, the endpoint will be added there.
```

Comment:
> Why not? I think the argument given here is that the endpoint needs to have its IP because you want to refer to it   elsewhere, regardless of whether it's attached to a container or not. But does it make any sense to refer to an IP that isn't attached to a container, or is attached to a stopped container?

I think it makes sense. Think of endpoint as the network part of a virtual machine and the container as the application running in it. You definitely want to assign network resources when the virtual machine spins up and not during application launch. This abstraction in virtual machine provided the convenience of swapping out different versions of the same application without the clients of the applications knowing it. The only difference is you can ping to a VM even without an application while you can't do that with an endpoint. But it's not a very important difference IMO.

> Actually that's a bit of a distraction: the real issue is that there's no way, as a driver, to provide IP addresses only to running containers. The driver API would need hooks that are called when a container starts and stops, for that. This is a separate issue to whether an endpoint keeps the same IP address or not.

This is intentional. Just continuing with the endpoint analogy, endpoint is like a virtual machine pre-provisioning but Join launches the endpoint(VM) on a particular host during Join and we are telling the driver about this particular event. The VM has been pre-provisioned with a particular IP address and it doesn't make a lot of sense to ask the driver for network resources during Join. 

> By the way, where are the requirements that motivate this design? I haven't seen anything suggesting a "service abstraction" before.

This is a feature specifically requested by Solomon and we (along with Solomon) brainstormed about the use cases and in the end we all agreed that this is a good abstraction to have. Overall of as I mentioned previously for implicitly created endpoints there is absolutely no difference where the network resources are acquired because docker core will make sure that `CreateEndpoint` and `Join` are called together. And it provides a nice composable api to facilitate advanced use cases where you want to swap out containers. So not sure why there is concern.